core/services/postgres/event_broadcaster.go

```diff
@@ -0,0 +1,326 @@
+package postgres
+
+import (
+	"context"
+	"database/sql"
+	"sync"
+	"time"
+
+	"github.com/lib/pq"
+	"github.com/pkg/errors"
+	"go.uber.org/multierr"
+
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+//go:generate mockery --name EventBroadcaster --output ./mocks/ --case=underscore
+//go:generate mockery --name Subscription --output ./mocks/ --case=underscore
+
+// EventBroadcaster opaquely manages a collection of Postgres event listeners
+// and broadcasts events to subscribers (with an optional payload filter).
+type EventBroadcaster interface {
+	Start() error
+	Stop() error
+	Subscribe(channel, payloadFilter string) (Subscription, error)
+	Notify(channel string, payload interface{}) error
+}
+
+type eventBroadcaster struct {
+	uri                  string
+	minReconnectInterval time.Duration
+	maxReconnectDuration time.Duration
+	db                   *sql.DB
+
+	listeners   map[string]*channelListener
+	listenersMu sync.Mutex
+
+	utils.StartStopOnce
+}
+
+var _ EventBroadcaster = (*eventBroadcaster)(nil)
+
+type Event struct {
+	Channel string
+	Payload string
+}
+
+func NewEventBroadcaster(uri string, minReconnectInterval time.Duration, maxReconnectDuration time.Duration) *eventBroadcaster {
+	if minReconnectInterval == time.Duration(0) {
+		minReconnectInterval = 1 * time.Second
+	}
+	if maxReconnectDuration == time.Duration(0) {
+		maxReconnectDuration = 1 * time.Minute
+	}
+	return &eventBroadcaster{
+		uri:                  uri,
+		minReconnectInterval: minReconnectInterval,
+		maxReconnectDuration: maxReconnectDuration,
+		listeners:            make(map[string]*channelListener),
+	}
+}
+
+func (b *eventBroadcaster) Start() error {
+	if !b.OkayToStart() {
+		return errors.Errorf("Postgres event broadcaster has already been started")
+	}
+	db, err := sql.Open("postgres", b.uri)
+	if err != nil {
+		return err
+	}
+	b.db = db
+	return nil
+}
+
+func (b *eventBroadcaster) Stop() error {
+	if !b.OkayToStop() {
+		return errors.Errorf("Postgres event broadcaster has already been stopped")
+	}
+
+	err := b.db.Close()
+
+	b.listenersMu.Lock()
+	defer b.listenersMu.Unlock()
+
+	for _, listener := range b.listeners {
+		err = multierr.Append(err, listener.stop())
+	}
+	b.listeners = nil // avoid "close of closed channel" panic on shutdown
+	return err
+}
+
+func (b *eventBroadcaster) Notify(channel string, payload interface{}) error {
```

Comment:
This is convenient, but the problem is we may wish to atomically generate notification events inside a transaction. In that case, the NOTIFY must be issued on the parent DB connection - not the eventBroadcaster's db connection.